### PR TITLE
Move various thread-local objects that were used in self-contained places out of Thread

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Move the following type's thread-specific trackers out of the `Thread` object and into self-contained fields/thread_local globals:

- StackingAllocator
- LoadLevelLimiter
- PendingTypeLoadHolder
- CLRRandom

Each of these were only used in one centralized location and the only usage of the `Thread` object was to fetch the object for the given thread. I've validated that for each of these cases, MSVC makes the same inlining decisions it made before this PR.

Also, remove the thread-specific RCW active list as that code is never used.

### CODE REVIEW

**Comment by EgorBo** [comment]
Does it address https://github.com/dotnet/runtime/issues/99661 too? `alloc_context` doesn't need to be inside Thread.

**Comment by jkoritzinsky** [comment]
It doesn't currently (as that's used a little more often than the metric I used), but I can try making it so.

File: src/coreclr/vm/appdomain.hpp [hunk 1]
```diff
@@ -822,6 +821,7 @@ typedef FileLoadLock::Holder FileLoadLockHolder;
 #endif
 class LoadLevelLimiter
 {
+    static thread_local LoadLevelLimiter* t_currentLoadLevelLimiter;
```
Comment (jkotas) [inline src/coreclr/vm/appdomain.hpp:824]: Is this only used for asserts? Should it be under `#ifdef _DEBUG`?

File: src/coreclr/vm/appdomain.hpp [hunk 2]
```diff
@@ -822,6 +821,7 @@ typedef FileLoadLock::Holder FileLoadLockHolder;
 #endif
```
Comment (jkotas) [inline src/coreclr/vm/appdomain.hpp:821]: The `#pragma warning (disable: 4324)` looks misplaced. Can it be deleted?

File: src/coreclr/vm/util.cpp [hunk 3]
```diff
@@ -1774,24 +1774,14 @@ static BOOL TrustMeIAmSafe(void *pLock)
 
 LockOwner g_lockTrustMeIAmThreadSafe = { NULL, TrustMeIAmSafe };
 
-static DangerousNonHostedSpinLock g_randomLock;
-static CLRRandom g_random;
+static thread_local CLRRandom t_random;
 
 int GetRandomInt(int maxVal)
```
Comment (jkotas) [inline src/coreclr/vm/util.cpp:1779]: There is random `GetRandomInt` in src\coreclr\gc\env\gcenv.base.h that just uses `rand()`. Can it be deleted?

File: src/coreclr/vm/util.cpp [hunk 4]
```diff
@@ -1774,24 +1774,14 @@ static BOOL TrustMeIAmSafe(void *pLock)
 
 LockOwner g_lockTrustMeIAmThreadSafe = { NULL, TrustMeIAmSafe };
 
-static DangerousNonHostedSpinLock g_randomLock;
```
Comment (jkotas) [inline src/coreclr/vm/util.cpp:1777]: Nearly all places that use this random number generator seems to be debug-only code, except one in ceemain.cpp; `ExecutableAllocator::InitLazyPreferredRange`. It seems wasteful to carry 200+ bytes of thread statics for each thread just to make that one place work in release builds.

**Comment by jkotas** [comment]
> Does it address https://github.com/dotnet/runtime/issues/99661 too? alloc_context doesn't need to be inside Thread.

This should be done in a separate change. It will chance a lot of places like assembly helpers.

File: src/coreclr/vm/appdomain.hpp (hunk 1 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/appdomain.hpp:824]: Looks like it is assert-only. I'll adjust it to be debug-only

Comment (jkoritzinsky) [reply src/coreclr/vm/appdomain.hpp:824]: Actually, it's used in one place, `AppDomain::LoadDomainAssembly`, so we can't make it debug-only.

Comment (jkotas) [reply src/coreclr/vm/appdomain.hpp:819]: Where is it exactly? I see it only in `POSTCONDITION` that is debug-only assert.

Comment (jkoritzinsky) [reply src/coreclr/vm/appdomain.hpp:819]: It's used here:

https://github.com/dotnet/runtime/blob/4d44e37d09a1cbff7c91a956d7da9ed980f9a4e5/src/coreclr/vm/appdomain.cpp#L2848-L2854

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by jkoritzinsky** [comment]
/ba-g The unrecognized failure is https://github.com/dotnet/runtime/issues/99142, which was closed as a fix was made after this PR started running CI

**Pull request merged**